GLTF2BIM/GLTF/GLTFBuilder.cs

```diff
@@ -315,12 +315,93 @@ public sealed partial class GLTFBuilder {
                 _primQueue.Clear();
             }
         }
+
+        private int SearchMesh(List<glTFMeshPrimitive> primitives)
+        {
+            var key = GetPrimitivesKey(primitives);
+
+            int meshIdx = default;
+
+            if (meshesInstancing.TryGetValue(key, out meshIdx))
+                return meshIdx;
+
+            return -1;
+        }
+        private uint? SearchMaterial(string name, float[] color)
+        {
+            var key = GetMaterialKey(name, color);
+
+            uint materialIdx = default;
+
+            if (materialsInstancing.TryGetValue(key, out materialIdx))
+                return materialIdx;
+
+            return null; 
+        }
+
+        private int CreateMesh()
+        {
+            var primitives = _primQueue.ToList();
+
+            var mesh = new glTFMesh
+            {
+                Primitives = primitives
+            };
+
+            // create a new mesh
+            _gltf.Meshes.Add(mesh);
+            var meshIdx = _gltf.Meshes.Count - 1;
+            meshesInstancing.Add(GetPrimitivesKey(primitives), meshIdx);
+
+            return meshIdx;
+        }
+        private uint CreateMaterial(string name, float[] color, glTFExtension[] exts, glTFExtras extras)
+        {
+            // it is a new material, proceed to add it!
+            var material = new glTFMaterial()
+            {
+                Name = name,
+                PBRMetallicRoughness = new glTFPBRMetallicRoughness()
+                {
+                    BaseColorFactor = color,
+                    MetallicFactor = 0f,
+                    RoughnessFactor = 1f,
+                },
+                Extensions = exts?.ToDictionary(x => x.Name, x => x),
+                Extras = extras
+            };
+
+            _gltf.Materials.Add(material);
+            var materialIdx = (uint)_gltf.Materials.Count - 1;
+            materialsInstancing.Add(GetMaterialKey(name, color), materialIdx);
+
+            return materialIdx;
+        }
+
+
+        public string GetPrimitivesKey(List<glTFMeshPrimitive> primitives)
+        {
+            string key = string.Empty;
+
+            foreach (var primitive in primitives)

```

Comment:
uses StringBuilder will be better in terms of performance

Reply:
thank you! just fixed it